### ISSUE TITLE
Your loss function seems problematic

### ISSUE THREAD

**Issue by kunrenzhilu** [tdeboissiere/DeepLearningImplementations#35]
It seems that you minimize the disc_loss_real first and then minimize the disc_loss_gen, however, according to the original paper, you should do the optimization of both simultaneously. Since this allow the optimizer to find the gradient of whole function in a global space, instead of following either the direction of the loss_real or loss_gen.

**Comment by tdeboissiere** [comment]
Which loss function (which GAN) are you referring to ?

**Comment by kunrenzhilu** [comment]
Oops, sorry I thought the issues is under WassersteinGAN already.

**Comment by tdeboissiere** [comment]
As far as I can tell, the authors are implementing the optimization for the discriminator in 2 steps as well. I have tried both and got much better convergence with the 2 steps approach.

**Comment by kunrenzhilu** [comment]
Sorry I see the reason. Since when updating the netD, the loss function is a linear combination of disc_loss_real and disc_loss_gen, and therefore the gradient can be computed separately. However, if you look at the @martinarjovsky 's code, you will find (line 199), the optimisation is updated in a global scale after combining two parts. However, in your code, you are really updating the weights separately, right?

**Comment by tdeboissiere** [comment]
Fair point, my explanation was off.

I think the issue here is Keras' API.
If I stacked X_real and X_gen together it is likely that the model would not train well as BatchNormarlization would be perturbed (see Soumith GAN tricks).
So I separated the updates to the discriminator in two distinct parts and it trained OK.

To correctly mimic the original code, I would have had to use more involved Keras hacking which I'd rather avoid.

**Comment by jiangzidong** [comment]
@tdeboissiere Can you please elaborate more about `Stacking X_real and X_gen together does not match the original algorithm and perturbs BatchNormarlization (see Soumith GAN tricks).`? 
* Which tricks in https://github.com/soumith/ganhacks
* Why not match original algorithm

Thanks alot
